### PR TITLE
Make miri a subtree instead of a submodule

### LINKED ISSUE #100134: Rust 1.62.1 cannot be compiled and installed, because of miri
When compiling rust 1.62.1 it fails with:
```
Building stage1 tool miri (x86_64-unknown-linux-gnu)
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /src/rustc-1.62.1-src/src/tools/rls/racer/Cargo.toml
workspace: /src/rustc-1.62.1-src/Cargo.toml
   Compiling miri v0.1.0 (/src/rustc-1.62.1-src/src/tools/miri)
error[E0407]: method `ptr_from_addr` is not a member of trait `Machine`
   --> src/tools/miri/src/machine.rs:602:5
    |
602 | /     fn ptr_from_addr(
603 | |         ecx: &MiriEvalContext<'mir, 'tcx>,
604 | |         addr: u64,
605 | |     ) -> Pointer<Option<Self::PointerTag>> {
606 | |         intptrcast::GlobalStateInner::ptr_from_addr(addr, ecx)
607 | |     }
    | |_____^ not a member of trait `Machine`

error[E0053]: method `get_alloc_id` has an incompatible type for trait
   --> src/tools/miri/src/machine.rs:161:30
    |
161 |     fn get_alloc_id(self) -> AllocId {
    |                              ^^^^^^^
    |                              |
    |                              expected enum `Option`, found struct `rustc_const_eval::interpret::AllocId`
    |                              help: change the output type to match the trait: `Option<rustc_const_eval::interpret::AllocId>`
    |
    = note: expected fn pointer `fn(machine::Tag) -> Option<rustc_const_eval::interpret::AllocId>`
               found fn pointer `fn(machine::Tag) -> rustc_const_eval::interpret::AllocId`

error[E0053]: method `ptr_get_alloc` has an incompatible type for trait
   --> src/tools/miri/src/machine.rs:614:10
    |
614 |     ) -> (AllocId, Size, Self::TagExtra) {
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |          |
    |          expected enum `Option`, found tuple
    |          help: change the output type to match the trait: `Option<(rustc_const_eval::interpret::AllocId, rustc_target::abi::Size, stacked_borrows::SbTag)>`
    |
    = note: expected fn pointer `fn(&rustc_const_eval::interpret::InterpCx<'mir, 'tcx, machine::Evaluator<'mir, 'tcx>>, rustc_const_eval::interpret::Pointer<_>) -> Option<(rustc_const_eval::interpret::AllocId, rustc_target::abi::Size, stacked_borrows::SbTag)>`
               found fn pointer `fn(&rustc_const_eval::interpret::InterpCx<'mir, 'tcx, machine::Evaluator<'mir, 'tcx>>, rustc_const_eval::interpret::Pointer<_>) -> (rustc_const_eval::interpret::AllocId, rustc_target::abi::Size, stacked_borrows::SbTag)`

error[E0046]: not all trait items implemented, missing: `ptr_from_addr_cast`, `ptr_from_addr_transmute`, `expose_ptr`
   --> src/tools/miri/src/machine.rs:434:1
    |
434 | impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'mir, 'tcx> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `ptr_from_addr_cast`, `ptr_from_addr_transmute`, `expose_ptr` in implementation
    |
    = help: implement the missing item: `fn ptr_from_addr_cast(_: &rustc_const_eval::interpret::InterpCx<'mir, 'tcx, Self>, _: u64) -> rustc_const_eval::interpret::Pointer<Option<<Self as rustc_const_eval::interpret::Machine<'mir, 'tcx>>::PointerTag>> { todo!() }`
    = help: implement the missing item: `fn ptr_from_addr_transmute(_: &rustc_const_eval::interpret::InterpCx<'mir, 'tcx, Self>, _: u64) -> rustc_const_eval::interpret::Pointer<Option<<Self as rustc_const_eval::interpret::Machine<'mir, 'tcx>>::PointerTag>> { todo!() }`
    = help: implement the missing item: `fn expose_ptr(_: &mut rustc_const_eval::interpret::InterpCx<'mir, 'tcx, Self>, _: rustc_const_eval::interpret::Pointer<<Self as rustc_const_eval::interpret::Machine<'mir, 'tcx>>::PointerTag>) -> Result<(), rustc_const_eval::interpret::InterpErrorInfo<'tcx>> { todo!() }`

error[E0061]: this function takes 2 arguments but 1 argument was supplied
    --> src/tools/miri/src/machine.rs:546:29
     |
546  |         let attrs = ecx.tcx.get_attrs(def_id);
     |                             ^^^^^^^^^ ------ supplied 1 argument
     |                             |
     |                             expected 2 arguments
     |
note: associated function defined here
    --> /src/rustc-1.62.1-src/compiler/rustc_middle/src/ty/mod.rs:2194:12
     |
2194 |     pub fn get_attrs(self, did: DefId, attr: Symbol) -> ty::Attributes<'tcx> {
     |            ^^^^^^^^^

error[E0308]: mismatched types
    --> src/tools/miri/src/machine.rs:547:73
     |
547  |         let link_name = match ecx.tcx.sess.first_attr_value_str_by_name(attrs, sym::link_name) {
     |                                                                         ^^^^^ expected `&[Attribute]`, found opaque type
    ::: /src/rustc-1.62.1-src/compiler/rustc_middle/src/ty/mod.rs:1944:29
     |
1944 | pub type Attributes<'tcx> = impl Iterator<Item = &'tcx ast::Attribute>;
     |                             ------------------------------------------ the found opaque type
     |
     = note: expected reference `&[Attribute]`
              found opaque type `rustc_middle::ty::Attributes<'_>`

error[E0061]: this function takes 2 arguments but 1 argument was supplied
    --> src/tools/miri/src/shims/foreign_items.rs:238:30
     |
238  |         let attrs = this.tcx.get_attrs(def_id);
     |                              ^^^^^^^^^ ------ supplied 1 argument
     |                              |
     |                              expected 2 arguments
     |
note: associated function defined here
    --> /src/rustc-1.62.1-src/compiler/rustc_middle/src/ty/mod.rs:2194:12
     |
2194 |     pub fn get_attrs(self, did: DefId, attr: Symbol) -> ty::Attributes<'tcx> {
     |            ^^^^^^^^^

error[E0308]: mismatched types
    --> src/tools/miri/src/shims/foreign_items.rs:242:43
     |
242  |             .first_attr_value_str_by_name(attrs, sym::link_name)
     |                                           ^^^^^ expected `&[Attribute]`, found opaque type
     |
    ::: /src/rustc-1.62.1-src/compiler/rustc_middle/src/ty/mod.rs:1944:29
     |
1944 | pub type Attributes<'tcx> = impl Iterator<Item = &'tcx ast::Attribute>;
     |                             ------------------------------------------ the found opaque type
     |
     = note: expected reference `&[Attribute]`
              found opaque type `rustc_middle::ty::Attributes<'_>`

error[E0599]: no method named `subst` found for struct `rustc_middle::ty::Ty` in the current scope
   --> src/tools/miri/src/shims/posix/sync.rs:844:21
    |
844 |     let ty = def_ty.subst(*tcx, &[param.into()]);
    |                     ^^^^^ method not found in `rustc_middle::ty::Ty<'_>`

warning: unused import: `subst::Subst`
 --> src/tools/miri/src/shims/posix/sync.rs:4:62
  |
4 | use rustc_middle::ty::{layout::TyAndLayout, query::TyCtxtAt, subst::Subst, Ty};
  |                                                              ^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

Some errors have detailed explanations: E0046, E0053, E0061, E0308, E0407, E0599.
For more information about an error, try `rustc --explain E0046`.
warning: `miri` (lib) generated 1 warning
error: could not compile `miri` due to 9 previous errors; 1 warning emitted
Building stage1 tool cargo-miri (x86_64-unknown-linux-gnu)
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /src/rustc-1.62.1-src/src/tools/rls/racer/Cargo.toml
workspace: /src/rustc-1.62.1-src/Cargo.toml
    Finished release [optimized] target(s) in 0.28s
Build completed successfully in 0:00:16
```

The above is during `x.py build -j1`.  When I call afterwards `x.py install -j1` the output is:
```
Building stage2 tool miri (x86_64-unknown-linux-gnu)
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /src/rustc-1.62.1-src/src/tools/rls/racer/Cargo.toml
workspace: /src/rustc-1.62.1-src/Cargo.toml
   Compiling miri v0.1.0 (/src/rustc-1.62.1-src/src/tools/miri)
error[E0407]: method `ptr_from_addr` is not a member of trait `Machine`
   --> src/tools/miri/src/machine.rs:602:5
    |
602 | /     fn ptr_from_addr(
603 | |         ecx: &MiriEvalContext<'mir, 'tcx>,
604 | |         addr: u64,
605 | |     ) -> Pointer<Option<Self::PointerTag>> {
606 | |         intptrcast::GlobalStateInner::ptr_from_addr(addr, ecx)
607 | |     }
    | |_____^ not a member of trait `Machine`

error[E0053]: method `get_alloc_id` has an incompatible type for trait
   --> src/tools/miri/src/machine.rs:161:30
    |
161 |     fn get_alloc_id(self) -> AllocId {
    |                              ^^^^^^^
    |                              |
    |                              expected enum `Option`, found struct `rustc_const_eval::interpret::AllocId`
    |                              help: change the output type to match the trait: `Option<rustc_const_eval::interpret::AllocId>`
    |
    = note: expected fn pointer `fn(machine::Tag) -> Option<rustc_const_eval::interpret::AllocId>`
               found fn pointer `fn(machine::Tag) -> rustc_const_eval::interpret::AllocId`

error[E0053]: method `ptr_get_alloc` has an incompatible type for trait
   --> src/tools/miri/src/machine.rs:614:10
    |
614 |     ) -> (AllocId, Size, Self::TagExtra) {
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |          |
    |          expected enum `Option`, found tuple
    |          help: change the output type to match the trait: `Option<(rustc_const_eval::interpret::AllocId, rustc_target::abi::Size, stacked_borrows::SbTag)>`
    |
    = note: expected fn pointer `fn(&rustc_const_eval::interpret::InterpCx<'mir, 'tcx, machine::Evaluator<'mir, 'tcx>>, rustc_const_eval::interpret::Pointer<_>) -> Option<(rustc_const_eval::interpret::AllocId, rustc_target::abi::Size, stacked_borrows::SbTag)>`
               found fn pointer `fn(&rustc_const_eval::interpret::InterpCx<'mir, 'tcx, machine::Evaluator<'mir, 'tcx>>, rustc_const_eval::interpret::Pointer<_>) -> (rustc_const_eval::interpret::AllocId, rustc_target::abi::Size, stacked_borrows::SbTag)`

error[E0046]: not all trait items implemented, missing: `ptr_from_addr_cast`, `ptr_from_addr_transmute`, `expose_ptr`
   --> src/tools/miri/src/machine.rs:434:1
    |
434 | impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'mir, 'tcx> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `ptr_from_addr_cast`, `ptr_from_addr_transmute`, `expose_ptr` in implementation
    |
    = help: implement the missing item: `fn ptr_from_addr_cast(_: &rustc_const_eval::interpret::InterpCx<'mir, 'tcx, Self>, _: u64) -> rustc_const_eval::interpret::Pointer<Option<<Self as rustc_const_eval::interpret::Machine<'mir, 'tcx>>::PointerTag>> { todo!() }`
    = help: implement the missing item: `fn ptr_from_addr_transmute(_: &rustc_const_eval::interpret::InterpCx<'mir, 'tcx, Self>, _: u64) -> rustc_const_eval::interpret::Pointer<Option<<Self as rustc_const_eval::interpret::Machine<'mir, 'tcx>>::PointerTag>> { todo!() }`
    = help: implement the missing item: `fn expose_ptr(_: &mut rustc_const_eval::interpret::InterpCx<'mir, 'tcx, Self>, _: rustc_const_eval::interpret::Pointer<<Self as rustc_const_eval::interpret::Machine<'mir, 'tcx>>::PointerTag>) -> Result<(), rustc_const_eval::interpret::InterpErrorInfo<'tcx>> { todo!() }`

error[E0061]: this function takes 2 arguments but 1 argument was supplied
    --> src/tools/miri/src/machine.rs:546:29
     |
546  |         let attrs = ecx.tcx.get_attrs(def_id);
     |                             ^^^^^^^^^ ------ supplied 1 argument
     |                             |
     |                             expected 2 arguments
     |
note: associated function defined here
    --> /src/rustc-1.62.1-src/compiler/rustc_middle/src/ty/mod.rs:2194:12
     |
2194 |     pub fn get_attrs(self, did: DefId, attr: Symbol) -> ty::Attributes<'tcx> {
     |            ^^^^^^^^^

error[E0308]: mismatched types
    --> src/tools/miri/src/machine.rs:547:73
     |
547  |         let link_name = match ecx.tcx.sess.first_attr_value_str_by_name(attrs, sym::link_name) {
     |                                                                         ^^^^^ expected `&[Attribute]`, found opaque type
     |
    ::: /src/rustc-1.62.1-src/compiler/rustc_middle/src/ty/mod.rs:1944:29
     |
1944 | pub type Attributes<'tcx> = impl Iterator<Item = &'tcx ast::Attribute>;
     |                             ------------------------------------------ the found opaque type
     |
     = note: expected reference `&[Attribute]`
              found opaque type `rustc_middle::ty::Attributes<'_>`

error[E0061]: this function takes 2 arguments but 1 argument was supplied
    --> src/tools/miri/src/shims/foreign_items.rs:238:30
     |
238  |         let attrs = this.tcx.get_attrs(def_id);
     |                              ^^^^^^^^^ ------ supplied 1 argument
     |                              |
     |                              expected 2 arguments
     |
note: associated function defined here
    --> /src/rustc-1.62.1-src/compiler/rustc_middle/src/ty/mod.rs:2194:12
     |
2194 |     pub fn get_attrs(self, did: DefId, attr: Symbol) -> ty::Attributes<'tcx> {
     |            ^^^^^^^^^

error[E0308]: mismatched types
    --> src/tools/miri/src/shims/foreign_items.rs:242:43
     |
242  |             .first_attr_value_str_by_name(attrs, sym::link_name)
     |                                           ^^^^^ expected `&[Attribute]`, found opaque type
     |
    ::: /src/rustc-1.62.1-src/compiler/rustc_middle/src/ty/mod.rs:1944:29
     |
1944 | pub type Attributes<'tcx> = impl Iterator<Item = &'tcx ast::Attribute>;
     |                             ------------------------------------------ the found opaque type
     |
     = note: expected reference `&[Attribute]`
              found opaque type `rustc_middle::ty::Attributes<'_>`

error[E0599]: no method named `subst` found for struct `rustc_middle::ty::Ty` in the current scope
   --> src/tools/miri/src/shims/posix/sync.rs:844:21
    |
844 |     let ty = def_ty.subst(*tcx, &[param.into()]);
    |                     ^^^^^ method not found in `rustc_middle::ty::Ty<'_>`

warning: unused import: `subst::Subst`
 --> src/tools/miri/src/shims/posix/sync.rs:4:62
  |
4 | use rustc_middle::ty::{layout::TyAndLayout, query::TyCtxtAt, subst::Subst, Ty};
  |                                                              ^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

Some errors have detailed explanations: E0046, E0053, E0061, E0308, E0407, E0599.
For more information about an error, try `rustc --explain E0046`.
warning: `miri` (lib) generated 1 warning
error: could not compile `miri` due to 9 previous errors; 1 warning emitted
thread 'main' panicked at 'Unable to build miri', src/bootstrap/dist.rs:42:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Build completed unsuccessfully in 0:42:12
```
That is, rust 1.62.1 cannot be compiled and installed.

### LINKED ISSUE #101867: `miri` no longer builds after rust-lang/rust#101410
Hello, this is your friendly neighborhood mergebot.
After merging PR rust-lang/rust#101410, I observed that the tool miri has failing tests.
A follow-up PR to the repository https://github.com/rust-lang/miri is needed to fix the fallout.

cc @dingxiangfei2009, do you think you would have time to do the follow-up work?
If so, that would be great!

### PR DESCRIPTION
**Pull request by oli-obk**
r? @RalfJung 

fixes #101867
fixes https://github.com/rust-lang/rust/issues/100134

### CODE REVIEW

File: src/tools/miri/git-subtree.sh [hunk 1]
```diff
@@ -0,0 +1,1024 @@
+#!/bin/sh
+#
+# git-subtree.sh: split/join git repositories in subdirectories of this one
```
Comment (jyn514) [inline src/tools/miri/git-subtree.sh:3]: This isn't specific to miri; let's put it in `src/etc` instead.

File: src/bootstrap/dist.rs [hunk 2]
```diff
@@ -1451,7 +1430,7 @@ impl Step for Extended {
 
         let xform = |p: &Path| {
             let mut contents = t!(fs::read_to_string(p));
-            for tool in &["rust-demangler", "rust-analyzer", "miri", "rustfmt"] {
+            for tool in &["rust-demangler", "rust-analyzer", "rustfmt"] {
```
Comment (jyn514) [inline src/bootstrap/dist.rs:1433]: Can you go ahead and remove `rust-analyzer` and `rustfmt` while you're at it? they're both subtrees now too :)

File: src/bootstrap/dist.rs [hunk 3]
```diff
@@ -1491,7 +1470,8 @@ impl Step for Extended {
             prepare("rust-std");
             prepare("rust-analysis");
             prepare("clippy");
-            for tool in &["rust-docs", "rust-demangler", "rust-analyzer", "miri"] {
+            prepare("miri");
+            for tool in &["rust-docs", "rust-demangler", "rust-analyzer"] {
```
Comment (jyn514) [inline src/bootstrap/dist.rs:1474]: ditto
```suggestion
            for tool in &["rust-docs", "rust-demangler"] {
```

File: src/bootstrap/dist.rs [hunk 4]
```diff
@@ -1550,7 +1530,8 @@ impl Step for Extended {
             prepare("rust-docs");
             prepare("rust-std");
             prepare("clippy");
-            for tool in &["rust-demangler", "rust-analyzer", "miri"] {
+            prepare("miri");
+            for tool in &["rust-demangler", "rust-analyzer"] {
```
Comment (jyn514) [inline src/bootstrap/dist.rs:1534]: ditto; and also on line 1618 below.

File: src/bootstrap/doc.rs [hunk 5]
```diff
@@ -851,7 +851,7 @@ tool_doc!(
     in_tree = true
 );
 tool_doc!(Clippy, "clippy", "src/tools/clippy", ["clippy_utils"], in_tree = true);
-tool_doc!(Miri, "miri", "src/tools/miri", ["miri"], in_tree = false);
+tool_doc!(Miri, "miri", "src/tools/miri", ["miri"], in_tree = true);
```
Comment (jyn514) [inline src/bootstrap/doc.rs:854]: looks like these are all in-tree now :) so we can remove the option and hard-code `true` in the macro

**Review by jyn514 — APPROVED** [review]
this is awesome :heart:

File: src/bootstrap/doc.rs (hunk 5 above)
Comment (oli-obk) [reply src/bootstrap/doc.rs:854]: I will do these cleanups in a follow up, getting the PR merged quickly is important as you cannot rebase a subtree addition

File: src/bootstrap/dist.rs (hunk 2 above)
Comment (oli-obk) [reply src/bootstrap/dist.rs:1433]: 🤦 yea, will do

**Comment by oli-obk** [comment]
@bors r-

```
error: failed to run custom build command for `libffi-sys v2.0.0`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools/release/build/libffi-sys-0c4808e136dd5138/build-script-build` (exit status: 101)
  --- stdout
  checking build system type... x86_64-pc-linux-gnu
  checking host system type... 
  --- stderr
  Invalid configuration `i686-pc-windows-gnu': Kernel `windows' not known to work with OS `gnu'.
  configure: error: /bin/bash ./config.sub i686-pc-windows-gnu failed
  thread 'main' panicked at 'Configuring libffi', /cargo/registry/src/github.com-1ecc6299db9ec823/libffi-sys-2.0.0/build/common.rs:8:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:02:19
== clock drift check ==
  local time: Tue Sep 20 07:13:47 UTC 2022
  network time: Tue, 20 Sep 2022 07:13:47 GMT
== end clock drift check ==
Error: Process completed with exit code 1.
```

File: src/etc/git-subtree.sh [hunk 6]
```diff
@@ -0,0 +1,1024 @@
+#!/bin/sh
+#
+# git-subtree.sh: split/join git repositories in subdirectories of this one
+#
+# Copyright (C) 2009 Avery Pennarun <apenwarr@gmail.com>
```
Comment (RalfJung) [inline src/etc/git-subtree.sh:5]: What is the license of this file? Does it need to be added to https://github.com/rust-lang/rust/blob/master/.reuse/dep5 ?

File: src/tools/build-manifest/src/main.rs [hunk 7]
```diff
@@ -261,29 +260,6 @@ impl Builder {
         t!(self.checksums.store_cache());
     }
 
-    /// If a tool does not pass its tests on *any* of Linux and Windows, don't ship
-    /// it on *all* targets, because tools like Miri can "cross-run" programs for
-    /// different targets, for example, run a program for `x86_64-pc-windows-msvc`
-    /// on `x86_64-unknown-linux-gnu`.
-    /// Right now, we do this only for Miri.
-    fn check_toolstate(&mut self) {
-        for file in &["toolstates-linux.json", "toolstates-windows.json"] {
```
Comment (RalfJung) [inline src/tools/build-manifest/src/main.rs:270]: Technically this means we could now remove those JSON files from the distributed tarball again, but I guess there is no harm in keeping it.

File: src/etc/git-subtree.sh (hunk 6 above)
Comment (oli-obk) [reply src/etc/git-subtree.sh:5]: it's the original git license afaict.

**Comment by RalfJung** [comment]
```
error: Tool `miri` was not recorded in tool state.
```

This line needs to be removed:

https://github.com/rust-lang/rust/blob/4a7e2fbb7bfdb7ddcb65b6fcf23bb5b272e66d42/src/bootstrap/toolstate.rs#L80

**Comment by RalfJung** [comment]
```
error: the feature `let_else` has been stable since 1.66.0-nightly and no longer requires an attribute to enable
  --> src/tools/miri/src/lib.rs:13:12
   |
13 | #![feature(let_else)]
   |            ^^^^^^^^
   |
   = note: `-D stable-features` implied by `-D warnings`
```
Looks like it needs the latest patches from the Miri repo -- I'll leave that to @oli-obk

**Comment by oli-obk** [comment]
oops, didn't remove the commit where I re-added that for --stage0

@bors r+

**Comment by RalfJung** [comment]
```
  Error: fatal: unable to access 'https://github.com/rust-lang-ci/rust/': The requested URL returned error: 429
  The process '/usr/local/bin/git' failed with exit code 128
```
@bors retry git fetch error 429

**Comment by RalfJung** [comment]
This will completely skew the commit stats since now old Miri commits are in the repo twice -- from the original Miri interpreter merge and from this submodule merge.^^

**Comment by oli-obk** [comment]
@bors retry fatal: unable to access 'https://github.com/rust-lang-ci/rust/': The requested URL returned error: 429

What is going on here? Did all the new commits cause CI to break?

**Comment by RalfJung** [comment]
I first thought it would be a general infra problem but only this PR seems to be affected. Very strange...

**Comment by oli-obk** [comment]
Ok, this failure is now real:

```
   readelf: Error: Not an ELF file - it has the wrong magic bytes at the start
  ../src/x86/ffi.c: In function 'ffi_prep_closure_loc':
  ../src/x86/ffi.c:612:1: warning: label 'out' defined but not used [-Wunused-label]
   out:
   ^~~
  ../src/x86/sysv.S: Assembler messages:
  ../src/x86/sysv.S:929: Error: junk at end of line, first unrecognized character is `,'
  make[2]: *** [Makefile:1276: src/x86/sysv.lo] Error 1
  make[1]: *** [Makefile:1376: install-recursive] Error 1
  make: *** [Makefile:3110: install] Error 2
  thread 'main' panicked at 'Building libffi', C:\Users\runneradmin\.cargo\registry\src\github.com-1285ae84e5963aae\libffi-sys-2.0.0\build\common.rs:8:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

and 

```
readelf: Error: Not an ELF file - it has the wrong magic bytes at the start
  ../src/java_raw_api.c: In function 'ffi_java_raw_call':
  ../src/java_raw_api.c:317:3: warning: 'ffi_java_raw_to_ptrarray' is deprecated [-Wdeprecated-declarations]
     ffi_java_raw_to_ptrarray (cif, raw, avalue);
     ^~~~~~~~~~~~~~~~~~~~~~~~
  ../src/java_raw_api.c:76:1: note: declared here
   ffi_java_raw_to_ptrarray (ffi_cif *cif, ffi_java_raw *raw, void **args)
   ^~~~~~~~~~~~~~~~~~~~~~~~
  ../src/java_raw_api.c: In function 'ffi_java_translate_args':
  ../src/java_raw_api.c:328:3: warning: 'ffi_java_raw_size' is deprecated [-Wdeprecated-declarations]
     ffi_java_raw *raw = (ffi_java_raw*)alloca (ffi_java_raw_size (cif));
     ^~~~~~~~~~~~
  ../src/java_raw_api.c:45:1: note: declared here
   ffi_java_raw_size (ffi_cif *cif)
   ^~~~~~~~~~~~~~~~~
  ../src/java_raw_api.c:331:3: warning: 'ffi_java_ptrarray_to_raw' is deprecated [-Wdeprecated-declarations]
     ffi_java_ptrarray_to_raw (cif, avalue, raw);
     ^~~~~~~~~~~~~~~~~~~~~~~~
  ../src/java_raw_api.c:157:1: note: declared here
   ffi_java_ptrarray_to_raw (ffi_cif *cif, void **args, ffi_java_raw *raw)
   ^~~~~~~~~~~~~~~~~~~~~~~~
  ../src/java_raw_api.c: In function 'ffi_prep_java_raw_closure':
  ../src/java_raw_api.c:369:3: warning: 'ffi_prep_java_raw_closure_loc' is deprecated [-Wdeprecated-declarations]
     return ffi_prep_java_raw_closure_loc (cl, cif, fun, user_data, cl);
     ^~~~~~
  ../src/java_raw_api.c:337:1: note: declared here
   ffi_prep_java_raw_closure_loc (ffi_java_raw_closure* cl,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ../src/x86/ffi64.c: In function 'ffi_call_int':
  ../src/x86/ffi64.c:596:33: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
       reg_args->gpr[gprcount++] = (unsigned long) rvalue;
                                   ^
  ../src/x86/ffi64.c: In function 'ffi_prep_closure_loc':
  ../src/x86/ffi64.c:783:1: warning: label 'out' defined but not used [-Wunused-label]
   out:
   ^~~
  ../src/x86/unix64.S: Assembler messages:
  ../src/x86/unix64.S:542: Error: junk at end of line, first unrecognized character is `,'
  make[2]: *** [Makefile:1276: src/x86/unix64.lo] Error 1
  make[1]: *** [Makefile:1376: install-recursive] Error 1
  make: *** [Makefile:3110: install] Error 2
  thread 'main' panicked at 'Building libffi', C:\Users\runneradmin\.cargo\registry\src\github.com-1ecc6299db9ec823\libffi-sys-2.0.0\build\common.rs:8:5
```

Maybe we should cfg out the libffi stuff to only have the dep on miri-supported platforms (x86_64 linux for a start)

**Comment by oli-obk** [comment]
Not supporting mingw seems ok ^^ everything else seems to fail. I'll open an issue for libffi though

**Comment by oli-obk** [comment]
Wait hold up, why didn't we see this before. Maybe we should just not ship miri on mingw?

**Comment by RalfJung** [comment]
Yeah I am surprised, we the FFI stuff worked on Windows before I thought?

Anyway, fine with making it Linux-only for now. (Please make a Miri PR for that though, it's hard to review Miri changes here.)

**Comment by oli-obk** [comment]
done: https://github.com/rust-lang/miri/pull/2555

**Comment by RalfJung** [comment]
uh...?
```
 Documenting miri v0.1.0 (/checkout/src/tools/miri)
error[E0275]: overflow evaluating the requirement `rustc_ast::token::TokenKind: std::marker::Send`
[...]
error: could not document `miri`
```

**Comment by oli-obk** [comment]
Lol have we ever tried documenting miri?

**Comment by saethlin** [comment]
Yes, when Miri builds its docs have been available on https://doc.rust-lang.org/nightly-rustc

Which is a relatively recent thing, I'm told. The docs did build recently.

**Comment by RalfJung** [comment]
Yeah the docs currently work (well they did before the most recent toolstate breakage).

We can revert https://github.com/rust-lang/rust/pull/98764 for now if it helps.

**Comment by jyn514** [comment]
I expect raising the type length limit like it suggests would work fine too.

**Comment by RalfJung** [comment]
(that's a very old log, the bot just a long time before posting it)

**Comment by RalfJung** [comment]
> I expect raising the type length limit like it suggests would work fine too.

But why would this start being a problem now?

@oli-obk should we do a regular submodule update first, both to unblock the distributed Miri and to make sure that this PR is a no-op for the Miri folder (so all build failures must definitely be due to the subtree switch)?

**Comment by oli-obk** [comment]
I just disabled the (newly enabled in this PR) check that docs actually build and we can fix them later

@bors r+

**Comment by oli-obk** [comment]
But yes, if this one doesn't go through, let's do a submodule bump, I have enough experience manually "rebasing" (recreating) this PR now for that not to be annoying

**Comment by jyn514** [comment]
@RalfJung @oli-obk would one of you mind making a follow-up PR to address the comments I had earlier?

**Comment by oli-obk** [comment]
Thanks for the reminder, should get to it tomorrow

File: src/ci/docker/host-x86_64/x86_64-gnu-tools/checktools.sh [hunk 8]
```diff
@@ -23,3 +22,4 @@ cat /tmp/toolstate/toolstates.json
 python3 "$X_PY" test --stage 2 check-tools
 python3 "$X_PY" test --stage 2 src/tools/clippy
 python3 "$X_PY" test --stage 2 src/tools/rustfmt
+python3 "$X_PY" test --stage 2 src/tools/miri
```
Comment (RalfJung) [inline src/ci/docker/host-x86_64/x86_64-gnu-tools/checktools.sh:25]: Is it possible that this only gave us back test coverage on Linux, but not on Windows? That's a problem...

**Pull request merged**